### PR TITLE
CFE-3470: Changed return type of and(), not(), or() to boolean (context)

### PR DESCRIPTION
**Pull request by olehermanse**
Let's see where this blows up.

### CODE REVIEW

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5829/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5829/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5829/

**Comment by nickanderson** [comment]
Wht is the jira ticket associated with this, or is there yet?

**Comment by olehermanse** [comment]
@nickanderson https://tracker.mender.io/browse/CFE-3470

**Review by vpodzime — DISMISSED** [review]
ACK

File: tests/acceptance/01_vars/02_functions/not.cf [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+body common control
+{
+    inputs => { "../../default.cf.sub" };
+    bundlesequence  => { default("$(this.promise_filename)") };
+    version => "1.0";
+}
+
+#######################################################
+
+bundle agent init
```
Comment (nickanderson) [inline tests/acceptance/01_vars/02_functions/not.cf:15]: Unless you are actually initing something, you don't need this empty bundle.

File: tests/acceptance/01_vars/02_functions/not.cf [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        "any",
+        not("!any"),
+        "cfengine",
+        not("!cfengine"),
+
+        # Variable expansion:
+        "$(true_variable)",
+        not("$(false_variable)"),
+        not(not("$(true_variable)")),
+        not(not(not("$(false_variable)"))),
```
Comment (nickanderson) [inline tests/acceptance/01_vars/02_functions/not.cf:41]: :exploding_head:

File: tests/acceptance/01_vars/02_functions/not.cf [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+        # Double expand:
+        not(not("$($(true_variable_name))")),
+        not(not(not("$($(false_variable_name))"))),
```
Comment (nickanderson) [inline tests/acceptance/01_vars/02_functions/not.cf:45]: What about triple expansion?

**Comment by olehermanse** [comment]
@cf-bottom now that I added acceptance test, please run in jenkins again

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5837/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5837/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5837/

**Review by nickanderson — APPROVED** [review]
A couple of minor comments but nothing that I would block the test for. Looks pretty good overall.

File: tests/acceptance/01_vars/02_functions/not.cf [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+#######################################################
+
+bundle agent check
+{
+
+  reports:
+    ok.(!fail)::
+      "$(this.promise_filename) Pass";
```
Comment (nickanderson) [inline tests/acceptance/01_vars/02_functions/not.cf:95]: if you use the dcs bundles then the test description will beemitted, and the test will be capable of being soft failed (not that it needs to be soft failed).

Comment (olehermanse) [reply tests/acceptance/01_vars/02_functions/not.cf:95]: @nickanderson :
```
R: test description: Test that not() behaves as expected
```

File: tests/acceptance/01_vars/02_functions/not.cf (hunk 3 above)
Comment (olehermanse) [reply tests/acceptance/01_vars/02_functions/not.cf:39]: Added.

Comment (olehermanse) [reply tests/acceptance/01_vars/02_functions/not.cf:39]: What about quadruple expansion?

File: tests/acceptance/01_vars/02_functions/not.cf (hunk 2 above)
Comment (olehermanse) [reply tests/acceptance/01_vars/02_functions/not.cf:35]: It's easy if you read the previous line, for every not you add, alternate true vs false ;)

File: tests/acceptance/01_vars/02_functions/not.cf (hunk 1 above)
Comment (olehermanse) [reply tests/acceptance/01_vars/02_functions/not.cf]: Thanks, removed.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5838/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5838/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5838/

**Review by nickanderson — APPROVED** [review]

**Comment by olehermanse** [comment]
Created this issue to track the failure in sequential-tests: https://tracker.mender.io/browse/ENT-6410

**Comment by olehermanse** [comment]
@olehermanse backport test only.

**Comment by olehermanse** [comment]
Cherry-pick of test only:
https://github.com/cfengine/core/pull/4373
https://github.com/cfengine/core/pull/4374

**Pull request merged**